crates/chia-datalayer/src/merkle/blob.rs

```diff
@@ -0,0 +1,2557 @@
+                }
+
+                let new_block = Block {
+                    metadata: NodeMetadata {
+                        node_type: NodeType::Internal,
+                        dirty: false,
+                    },
+                    node: Node::Internal(InternalNode {
+                        parent: Parent(None),
+                        hash: internal_hash(&hashes[0], &hashes[1]),
+                        left: index_1,
+                        right: index_2,
+                    }),
+                };
+
+                self.insert_entry_to_blob(new_internal_node_index, &new_block)?;
+                new_indexes.push(new_internal_node_index);
+            }
+
+            indexes = new_indexes;
+        }
+
+        if indexes.len() == 1 {
+            // OPT: can we avoid this extra min height leaf traversal?
+            let min_height_leaf = self.get_min_height_leaf()?;
+            self.insert_subtree_at_key(min_height_leaf.key, indexes[0], Side::Left)?;
+        };
+
+        Ok(())
+    }
+
+    fn insert_subtree_at_key(
+        &mut self,
+        old_leaf_key: KeyId,
+        new_index: TreeIndex,
+        side: Side,
+    ) -> Result<(), Error> {
+        // TODO: seems like this ought to be fairly similar to regular insert
+
+        // TODO: but what about the old leaf being the root...  is that what the batch insert
+        //       pre-filling of two leafs is about?  if so, this needs to be making sure of that
+        //       or something.
+
+        struct Stuff {
+            index: TreeIndex,
+            hash: Hash,
+        }
+
+        let new_internal_node_index = self.get_new_index();
+        let (old_leaf_index, old_leaf, _old_block) = self.get_leaf_by_key(old_leaf_key)?;
+        let new_node = self.get_node(new_index)?;
+
+        let new_stuff = Stuff {
+            index: new_index,
+            hash: new_node.hash(),
+        };
+        let old_stuff = Stuff {
+            index: old_leaf_index,
+            hash: old_leaf.hash,
+        };
+        let (left, right) = match side {
+            Side::Left => (new_stuff, old_stuff),
+            Side::Right => (old_stuff, new_stuff),
+        };
+        let internal_node_hash = internal_hash(&left.hash, &right.hash);
+
+        let block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Internal,
+                dirty: false,
+            },
+            node: Node::Internal(InternalNode {
+                parent: old_leaf.parent,
+                hash: internal_node_hash,
+                left: left.index,
+                right: right.index,
+            }),
+        };
+        self.insert_entry_to_blob(new_internal_node_index, &block)?;
+        self.update_parent(new_index, Some(new_internal_node_index))?;
+
+        // TODO: expect relates to comment at the beginning about assumptions about the tree etc
+        let old_leaf_parent = old_leaf.parent.0.expect("not handling this case");
```

Comment:
is it acceptable to panic here? the message looks like it's something that could happen